### PR TITLE
GH-139: Improve preflight config validation

### PR DESCRIPTION
**Pull request by C0urante**
Depends on https://github.com/confluentinc/kafka-connect-bigquery/pull/140 (diff can be viewed [here](https://github.com/C0urante/kafka-connect-bigquery/compare/gh-139...confluentinc:gh-139-2)).

Addresses the remainder of https://github.com/confluentinc/kafka-connect-bigquery/issues/139.

All multi-property configuration validation is moved from the constructor of the BigQuerySinkConfig class to a new `BigQuerySinkConfig::validate` method, which is invoked from `BigQuerySinkConnector::validate` and leverages the Connect API for defining custom error messages on a per-property basis instead of throwing exceptions on invalid configurations, which only displays one message at a time and doesn't work well with programmatic UIs.

Unit tests are added for all new validation logic.

The logic for GCP client construction is also updated to conform to a common interface, which should improve readability and maintainability, and makes it easier to validate credentials for either BigQuery or GCS.

Finally, the `SinkConfigConnectException` class is removed as it's not really necessary and doesn't bring anything worthwhile to the code base.

### CODE REVIEW

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java
Comment (SajanaW) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:339]: Do we have guarantees that this would cast correctly? AFAICT the framework hands a List string object because we set type to list. Wondering if we should be safe and wrap this in a try/catch

**Review by SajanaW — COMMENTED** [review]
Looked at a few files and left a couple of comments. Will come back to take another pass!

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkConnector.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkConnector.java:121]: Good call. Although Kafka Connect does preflight validation whenever a new connector config is submitted (and rejects the config if any errors are detected), that doesn't cover the opposite case: the same connector config is present in the config topic before and after an upgrade of the connector. If the connector is upgraded and no longer works with the old config, an extra validation here will help surface that.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java
Comment (ddasarathan) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:640]: Is there a reason `name` is removed from the definition?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/MultiPropertyValidator.java
Comment (ddasarathan) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/MultiPropertyValidator.java:60]: It does look like all the inherited validator classes handle for `RuntimeException` in `doValidate` method. Do we need this here?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/SchemaRetrieverValidator.java
Comment (ddasarathan) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/SchemaRetrieverValidator.java:60]: Neatly done

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/config/PartitioningModeValidatorTest.java
Comment (ddasarathan) [inline kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/config/PartitioningModeValidatorTest.java:37]: Can we also add a test where this is `true` but TimestampPatitionFieldName is not set?

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/TableExistenceValidator.java
Comment (ddasarathan) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/TableExistenceValidator.java:88]: This could throw `BigQueryException`. I now see why the parent class `RuntimeException` may be needed.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/CredentialsValidator.java
Comment (ddasarathan) [inline kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/CredentialsValidator.java:69]: Do we need this?

**Review by ddasarathan — APPROVED** [review]
Have some minor comments, but otherwise LGTM

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/BigQuerySinkConfig.java:640]: It makes the method cleaner (easier to use with method reference syntax, for example, like is done in the `MapValidator` class), and it's unnecessary since when the method is invoked in the `MapValidator`, any errors that come up from it will automatically be associated with the property that's being validated and displayed to the user in the REST API response for the request to create/reconfigure the connector or validate a config for it.

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/SchemaRetrieverValidator.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/SchemaRetrieverValidator.java:60]: ❤️

File: kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/CredentialsValidator.java
Comment (C0urante) [reply kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/config/CredentialsValidator.java:69]: Yeah fair, I wanted to avoid having a trailing `": "` at the end if `e.getMessage()` was null but this is much more complicated than it needs to be just to handle that case.

File: kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/config/PartitioningModeValidatorTest.java
Comment (C0urante) [reply kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/config/PartitioningModeValidatorTest.java:37]: Good call, added.

**Comment by C0urante** [comment]
Thanks for the review @ddasarathan! I hope I've addressed your comments; given the LGTM I'll merge but please let me know if there's any follow-up.

**Pull request merged**
